playground/knot/src/Editor/Editor.tsx

```diff
@@ -347,6 +359,33 @@ class PlaygroundServer
 
     return true;
   }
+
+  provideDocumentFormattingEdits(
+    model: editor.ITextModel,
+  ): languages.ProviderResult<languages.TextEdit[]> {
+    if (this.props.current.files.selected == null) {
+      return;
+    }
+
+    const fileHandle =
+      this.props.current.files.handles[this.props.current.files.selected];
+
+    if (fileHandle == null) {
+      return;
+    }
+
+    const formatted = this.props.current.workspace.format(fileHandle);
+    if (formatted !== model.getValue()) {
```

Comment:
We could be more fancy and change `format` to return a vec of text edits. Again, this seems nice to have but probably not something we need for the editor.